### PR TITLE
Added tmx animated tile support, with test case. 

### PR DESCRIPTION
**Pull request by DEAGS3000**
TileAnimations are disabled by default, if user didn't enable it, the TMXTiledMap class will do just the same as this support not exists, no additional render resources needed. The tile animations are still rendered through SpriteBatchNode, it won't increase the draw call.

### CODE REVIEW

**Comment by minggo** [comment]
Is it a new feature of tilemap editor?

**Comment by DEAGS3000** [comment]
> Is it a new feature of tilemap editor?

Not quite new. According to Tiled's change log, it began to support animated tiles since Tiled 0.10 and now its version is 1.2+.

File: cocos/2d/CCTMXLayer.cpp [hunk 1]
```diff
@@ -845,5 +854,91 @@ std::string TMXLayer::getDescription() const
...
+        return;
+    _currentFrame = (_currentFrame + 1) % _frameCount;
+    _layer->setTileGID(_animation->_frames[_currentFrame]._tileID, _tilePosition);
+    scheduleNextTick();
+}
+
+void TMXTileAnimTask::scheduleNextTick()
+{
+    _started = true;
+//    std::stringstream ss;
```
Comment (Mee-gu) [inline cocos/2d/CCTMXLayer.cpp]: unneeded codes can be removed.

File: cocos/2d/CCTMXXMLParser.cpp [hunk 2]
```diff
@@ -672,6 +672,21 @@ void TMXMapInfo::startElement(void* /*ctx*/, const char *name, const char **atts
             dict["polylinePoints"] = Value(pointsArray);
         }
     }
+    else if(elementName == "animation")
+    {
+        TMXTilesetInfo* info = tmxMapInfo->getTilesets().back();
+        info->_animationInfo.insert(tmxMapInfo->getParentGID(), TMXTileAnimInfo::create(tmxMapInfo->getParentGID()));
+        //tmxMapInfo->setParentGID(info->_firstGid + attributeDict["id"].asInt());
```
Comment (Mee-gu) [inline cocos/2d/CCTMXXMLParser.cpp]: same as before.

File: cocos/2d/CCTMXLayer.cpp (hunk 1 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.cpp]: Got it. Working on it

File: cocos/2d/CCTMXLayer.h [hunk 3]
```diff
@@ -294,6 +297,22 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
 
+    /** Map from gid of animated tile to its instance.
+     *
+     * @return Map from gid of animated tile to its instance.
+     */
+    const std::map<uint32_t, std::vector<Vec2>>* getAnimTileCoord() {
+        return &_animTileCoord;
+    }
+
+    bool hasTileAnimation() {
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: It is better to add const like 
```c++
bool hasTileAnimation() const {}
```

File: cocos/2d/CCTMXLayer.h [hunk 4]
```diff
@@ -294,6 +297,22 @@ class CC_DLL TMXLayer : public SpriteBatchNode
     */
     virtual std::string getDescription() const override;
 
+    /** Map from gid of animated tile to its instance.
+     *
+     * @return Map from gid of animated tile to its instance.
+     */
+    const std::map<uint32_t, std::vector<Vec2>>* getAnimTileCoord() {
+        return &_animTileCoord;
+    }
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h:304]: Why not just return const reference like 
```c++
const std::map<uint32_t, std::vector<Vec2>>& getAnimTileCoord() const
```

File: cocos/2d/CCTMXLayer.h [hunk 5]
```diff
@@ -294,6 +297,22 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+
+    bool hasTileAnimation() {
+        return !_animTileCoord.empty();
+    }
+
+    TMXTileAnimManager* getTileAnimManager() {
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: Need const.

File: cocos/2d/CCTMXLayer.h [hunk 6]
```diff
@@ -353,6 +372,65 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+    TMXTileAnimManager *_tileAnimManager = nullptr;
+};
+
+/** @brief TMXTileAnimTask represents the frame-tick task of an animated tile.
+ * It is a assistant class for TMXTileAnimTicker.
+ */
+class TMXTileAnimTask
+{
+public:
+    TMXTileAnimTask(TMXLayer *layer, TMXTileAnimInfo *animation, Vec2 &tilePos);
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: Vec2 &tilePos -> const Vec2 &tilePos

File: cocos/2d/CCTMXLayer.h [hunk 7]
```diff
@@ -353,6 +372,65 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+    /** start the animation task */
+    void start();
+    /** stop the animation task */
+    void stop();
+    bool isRunning() {
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: need const

File: cocos/2d/CCTMXLayer.h [hunk 8]
```diff
@@ -353,6 +372,65 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+    /** start all tile animations */
+    void startAll();
+    /** stop all tile animations */
+    void stopAll();
+
+
+    bool _started = false;
+    /** vector contains all tasks of this layer */
+    std::vector<TMXTileAnimTask> _tasks;
+    TMXLayer* _layer;
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: Why variables be public?

File: cocos/2d/CCTMXLayer.h [hunk 9]
```diff
@@ -353,6 +372,65 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+    /** key of schedule task for specific animated tile */
+    std::string _key;
+    TMXLayer *_layer;
+    /** position of the animated tile */
+    Vec2 _tilePosition;
+    /** AnimationInfo on this tile */
+    TMXTileAnimInfo *_animation;
+    /** Index of the frame that should be drawn currently */
+    uint32_t _currentFrame;
+    uint32_t _frameCount;
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: It is better to set default values.

File: cocos/2d/CCTMXXMLParser.cpp [hunk 10]
```diff
@@ -807,4 +824,30 @@ void TMXMapInfo::textHandler(void* /*ctx*/, const char *ch, size_t len)
     }
 }
 
+TMXTileAnimFrame::TMXTileAnimFrame(uint32_t tileID, float duration)
+{
+    _tileID = tileID;
+    _duration = duration;
```
Comment (minggo) [inline cocos/2d/CCTMXXMLParser.cpp]: Use [constructor member initializer lists](https://www.learncpp.com/cpp-tutorial/8-5a-constructor-member-initializer-lists/) is better.

File: cocos/2d/CCTMXXMLParser.cpp [hunk 11]
```diff
@@ -807,4 +824,30 @@ void TMXMapInfo::textHandler(void* /*ctx*/, const char *ch, size_t len)
...
+}
+
+TMXTileAnimInfo::TMXTileAnimInfo(uint32_t tileID)
+{
+    _tileID = tileID;
```
Comment (minggo) [inline cocos/2d/CCTMXXMLParser.cpp]: The same as above.

File: cocos/2d/CCTMXXMLParser.h [hunk 12]
```diff
@@ -119,6 +121,35 @@ class CC_DLL TMXLayerInfo : public Ref
...
+
+This information is obtained from the TMX file.
+*/
+struct TMXTileAnimFrame
+{
+    TMXTileAnimFrame(uint32_t tileID, float duration);
+    /** gid of the frame */
+    uint32_t _tileID;
+    /** duration of the frame */
+    float _duration;
```
Comment (minggo) [inline cocos/2d/CCTMXXMLParser.h]: It is better to set default values.

File: cocos/2d/CCTMXXMLParser.h [hunk 13]
```diff
@@ -119,6 +121,35 @@ class CC_DLL TMXLayerInfo : public Ref
...
+- frames the animated tile contains
+
+This information is obtained from the TMX file.
+*/
+struct TMXTileAnimInfo : public Ref
+{
+    static TMXTileAnimInfo * create(uint32_t  tileID);
+    explicit TMXTileAnimInfo(uint32_t  tileID);
+    uint32_t _tileID;
+    std::vector<TMXTileAnimFrame> _frames;
```
Comment (minggo) [inline cocos/2d/CCTMXXMLParser.h:150]: The same as above.

File: cocos/2d/CCTMXLayer.h [hunk 14]
```diff
@@ -36,8 +36,11 @@ NS_CC_BEGIN
 class TMXMapInfo;
 class TMXLayerInfo;
 class TMXTilesetInfo;
+class TMXTileAnimManager;
 struct _ccCArray;
 
+
+
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: Unneeded spaces.

File: cocos/2d/CCTMXLayer.h (hunk 4 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h:304]: I tried before, but it will cause some problom about converting to lua values.

Comment (minggo) [reply cocos/2d/CCTMXLayer.h:304]: Yep, it may have problem in lua.

Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h:304]: Like this:

`C:\projects\cocos2d-x\cocos\scripting\lua-bindings\auto\lua_cocos2dx_auto.cpp(101214): error C2770: invalid explicit template argument(s) for 'void object_to_luaval(lua_State *,const char *,T *)' [C:\projects\cocos2d-x\win32-build\engine\cocos\lua-bindings\luacocos2d.vcxproj]`

File: cocos/2d/CCTMXLayer.h (hunk 8 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: Got it, but about `_tasks`, if users want to control a particular tile animation, then we'll need a getter for it, I wonder if the lua value conversion error will appear again?

Comment (minggo) [reply cocos/2d/CCTMXLayer.h]: But now how to access `_tasks` in lua?

Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: Hmmm, you are right

**Comment by Mee-gu** [comment]
I use Tiled Map Editor with Version 1.2.4 but can't find the Animation Editor button.
How do you add or edit a tile animation?

**Comment by DEAGS3000** [comment]
> I use Tiled Map Editor with Version 1.2.4 but can't find the Animation Editor button.
> How do you add or edit a tile animation?

It's in tileset. Enter the tileset-editing window, and theres a camera button in toolbar.

**Comment by Mee-gu** [comment]
> > I use Tiled Map Editor with Version 1.2.4 but can't find the Animation Editor button.
> > How do you add or edit a tile animation?
> 
> It's in tileset. Enter the tileset-editing window, and theres a camera button in toolbar.

I can't find the camera button image in the toolbar

image

Which Tiled Map Editor version do you use?

**Comment by DEAGS3000** [comment]
> > > I use Tiled Map Editor with Version 1.2.4 but can't find the Animation Editor button.
> > > How do you add or edit a tile animation?
> > 
> > 
> > It's in tileset. Enter the tileset-editing window, and theres a camera button in toolbar.
> 
> I can't find the camera button image in the toolbar
> 
> image
> 
> Which Tiled Map Editor version do you use?

No，not this window. You should add a tileset, and push the edit button of tileset, then you'll enter a new window, the camera button is there.

My version is 1.2.5, but 1.2.4 definitely has that feature.

**Comment by DEAGS3000** [comment]
I tried to regenerate sln and vcxproj files with cmake, and they goes well on Windows, no LNK2019 anymore. But it's hard to find the particular difference between them that causes link error, the most obvious one is, those project file generated by cmake has a preprocess definition `CC_STATIC` and original project file dose not. I got a power outage here and can't test that for now. Do you think that definition causes LNK2019 error?

**Comment by Mee-gu** [comment]
> You should add a tileset, and push the edit button of tileset, then you'll enter a new window, the camera button is there.

- `You should add a tileset, ` do you means File > New > New Tileset ?
-  what do you mean by`push the edit button of tileset`? Could you show me a picture?

**Comment by DEAGS3000** [comment]
> * do you means File > New > New Tileset ?

image

In this window, new tileset(first button) and edit tileset(forth button) are all here at bottom toolbar.

**Comment by Mee-gu** [comment]
> > * do you means File > New > New Tileset ?
> 
> image
> 
> In this window, new tileset(first button) and edit tileset(forth button) are all here at bottom toolbar.

Got it, thank you!

File: cocos/2d/CCTMXLayer.cpp (hunk 1 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.cpp]: fixed.

File: cocos/2d/CCTMXXMLParser.cpp (hunk 2 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXXMLParser.cpp]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 3 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 5 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 6 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 7 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 8 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed. Turned to cocos2d::Vector.

File: cocos/2d/CCTMXLayer.h (hunk 9 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXXMLParser.cpp (hunk 10 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXXMLParser.cpp]: fixed.

File: cocos/2d/CCTMXXMLParser.cpp (hunk 11 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXXMLParser.cpp]: fixed.

File: cocos/2d/CCTMXXMLParser.h (hunk 12 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXXMLParser.h]: fixed.

File: cocos/2d/CCTMXXMLParser.h (hunk 13 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXXMLParser.h:150]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 14 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.h [hunk 15]
```diff
@@ -353,6 +370,72 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+     */
+    Vector<TMXTileAnimTask*>& getTasks(){
+        return _tasks;
+    }
+
+protected:
+    bool _started = false;
+    /** vector contains all tasks of this layer */
+    Vector<TMXTileAnimTask*> _tasks;
+    TMXLayer* _layer;
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h]: Should initialize default value.

File: cocos/2d/CCTMXLayer.h [hunk 16]
```diff
@@ -353,6 +370,72 @@ class CC_DLL TMXLayer : public SpriteBatchNode
...
+    /** start all tile animations */
+    void startAll();
+    /** stop all tile animations */
+    void stopAll();
+
+    /** get vector of tasks
+     */
+    Vector<TMXTileAnimTask*>& getTasks(){
+        return _tasks;
+    }
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.h:431]: It is better to add const.

File: cocos/2d/CCTMXLayer.cpp [hunk 17]
```diff
@@ -845,5 +854,105 @@ std::string TMXLayer::getDescription() const
...
+    CC_SAFE_DELETE(ret);
+    return nullptr;
+}
+
 
 NS_CC_END
+
+
+
+
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.cpp]: I think only one new line is needed.

File: cocos/2d/CCTMXLayer.h (hunk 16 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

File: cocos/2d/CCTMXLayer.cpp (hunk 17 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.cpp]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 16 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h:431]: fixed.

File: cocos/2d/CCTMXLayer.h (hunk 15 above)
Comment (DEAGS3000) [reply cocos/2d/CCTMXLayer.h]: fixed.

**Review by minggo — APPROVED** [review]

**Comment by Nordsoft91** [comment]
When will be merged into v4? I manually merge it into v4 branch and it works good.

**Pull request merged**